### PR TITLE
Fix expire triggers.

### LINKED ISSUE #5354: graph parser: expire-all ?
The `expire-all` trigger (valid at Cylc 7) appears to be missing from Cylc 8:

https://github.com/cylc/cylc-flow/blob/76762bbfc1ca83be88ed8b0e0a7224e6af520964/cylc/flow/graph_parser.py#L105-L116

I'm not sure whether `expire-all` was "officially supported" or just a side-effect of the way the logic worked, however, it seems like a sensible trigger condition to me. I'm not sure whether its removal was intentional, nothing springs to mind.

At Cylc 8 such a trigger would probably required a `?` modifier (as succeeded and expired are orthogonal), but should still work.

**Pull requests welcome!**

### LINKED ISSUE #5364: :expire trigger broken
By broken I mean, it has no effect.

When implementing the spawn-on-demand scheduler, I think I must have forgotten about the `:expired` trigger, and hence removed expired tasks before triggering on that "output" (because an expired task does not have any *other* outputs).

### LINKED ISSUE #5447: expiry: broken interaction with runahead limit
There is a bug in the interaction between runahead limit and task expiry.

In the following example, Cylc spawns the task out to the runahead limit, these tasks then clock-expire. The task pool is now empty so the workflow shuts down, job done, despite there still being un-expired tasks left to run before the FCP.

```
[scheduling]                                         
    initial cycle point = previous(T-00) - PT5H    
    final cycle point = +PT10H    
    runahead limit = P3                                                                                                            
    [[special tasks]]                                
        clock-expire = task1(PT0H)                   
    [[dependencies]]                                 
        [[[R//PT1H]]]                                
            graph = "@wall_clock => task1"           
                                                     
[runtime]                                            
    [[task1]]
```

### PR DESCRIPTION
**Pull request by hjoliver**
- fix expire trigger: close #5364 
- fix expire-all/any triggers: close #5354  
- fix #5447

And require `?` (optional) on all expire triggers.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/596
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
Converted back to draft temporarily, in light of today' project meeting.

I think have nailed the expire stall/no-stall situation, ran out of time to write it up this afternoon though.

File: cylc/flow/task_pool.py [hunk 1]
```diff
@@ -686,6 +686,7 @@ def remove(self, itask, reason=""):
             if not self.hidden_pool[itask.point]:
                 del self.hidden_pool[itask.point]
             LOG.debug(f"[{itask}] {msg}")
+            self.task_queue_mgr.remove_task(itask)
```
Comment (oliver-sanders) [inline cylc/flow/task_pool.py:689]: Unrelated bug?

Comment (hjoliver) [reply cylc/flow/task_pool.py:689]: It is a bug fix, but not unrelated. Expired tasks need to be removed as complete (possibly still subject to debate!) but the "remove" process was not removing the target task from the queues.

**Comment by hjoliver** [comment]
Note this PR is largely orthogonal to the additional and currently unresolved discussion of what we should do in response to an expiration event beyond just triggering off of it (e.g. stall as well if the graph doesn't explicitly handle the expiration). However it might make sense to bundle the result of that with this PR.

File: cylc/flow/task_pool.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/task_pool.py:689]: Having run into this bug again, I'll cherry-pick it to a separate bug fix PR. (As task expiry is a bigger discussion).

Comment (hjoliver) [reply cylc/flow/task_pool.py:689]: Having run into this bug again, I'll cherry-pick it to a separate bug fix PR. (As task expiry is a bigger discussion).... #5573

Comment (hjoliver) [reply cylc/flow/task_pool.py:689]: ... #5573

**Comment by hjoliver** [comment]
Closing as superseded by #5658

**Pull request closed without merging**